src/poetry/core/packages/constraints/union_constraint.py

```diff
@@ -99,22 +100,22 @@ def intersect(self, other: "ConstraintTypes") -> "ConstraintTypes":
 
         return UnionConstraint(*new_constraints)
 
-    def union(self, other: Constraint) -> "UnionConstraint":
-        if isinstance(other, Constraint):
-            constraints = self._constraints
-            if other not in self._constraints:
-                constraints += (other,)
+    def union(self, other: BaseConstraint) -> "UnionConstraint":
+        if not isinstance(other, Constraint):
+            raise ValueError("Unimplemented constraint union")
```

Comment:
If we only allow Constraint we should tell it so by the type hint.

Reply:
As elsewhere: this would be inconsistent with the base type which mypy would rightly object to.  Again, a `Constraint` is logically allowed here: but we have no implementation for that case.